### PR TITLE
Avoid classical for when looping through all items of an Array-like object

### PR DESCRIPTION
**Pull request by teoli2003**
Classical `for` loops are error-prone (setting the right condition, not mixing indices…) and difficult to read when used to loop on all elements of an Array-like structure.

This PR changes some occurrences to use the modern `for…of`, or `.forEach()` in some occurrences.

Note that:
- I use `for…of` if I don't need the index
- I use `.forEach()` when I do, or on an object where `for…of` may not work.
- There were a couple of cases where the loop copied an Array into a new one; I used `Array.from()` instead.

### CODE REVIEW

File: files/en-us/web/api/vrdisplaycapabilities/index.md [hunk 1]
```diff
@@ -39,20 +39,20 @@ This interface is accessible through the {{domxref("VRDisplay.capabilities")}} p
 ```js
 function reportDisplays() {
   navigator.getVRDisplays().then((displays) => {
-    for (let i = 0; i < displays.length; i++) {
-      const cap = displays[i].capabilities;
+    displays.foreach((display, i) => {
```
Comment (zloirock) [inline files/en-us/web/api/vrdisplaycapabilities/index.md:42]: Typo.

Comment (teoli2003) [reply files/en-us/web/api/vrdisplaycapabilities/index.md:42]: Oh yes, I do this one so often…
```suggestion
    displays.forEach((display, i) => {
```

File: files/en-us/web/api/webglrenderingcontext/vertexattribpointer/index.md [hunk 2]
```diff
@@ -218,13 +218,13 @@ First, we dynamically create the array buffer from JSON data using a
...
 const response = await fetch('assets/geometry.json');
 const vertices = await response.json();
 
-//Create array buffer
+// Create array buffer
 const buffer = new ArrayBuffer(20 * vertices.length);
-//Fill array buffer
+// Fill array buffer
 const dv = new DataView(buffer);
 for (let i = 0; i < vertices.length; i++) {
```
Comment (Josh-Cena) [inline files/en-us/web/api/webglrenderingcontext/vertexattribpointer/index.md:229]: This looks like a `forEach`?

File: files/en-us/web/api/webrtc_api/signaling_and_video_calling/index.md [hunk 3]
```diff
@@ -45,11 +45,10 @@ First up is the addition of the function `sendToOneUser()`. As the name suggests
...
 
-  for (i=0; i < connectionArray.length; i++) {
-    if (connectionArray[i].username === target) {
-      connectionArray[i].send(msgString);
+  for (const connection of connectionArray) {
+    if (connection.username === target) {
+      connection.send(msgString);
       break;
     }
   }
```
Comment (Josh-Cena) [inline files/en-us/web/api/webrtc_api/signaling_and_video_calling/index.md:54]:
```suggestion
  connectionArray.find((conn) => conn.username === target).send(msgString);
```

File: files/en-us/web/api/websockets_api/writing_websocket_servers/index.md [hunk 4]
```diff
@@ -128,18 +128,16 @@ To read the payload data, you must know when to stop reading. That's why the pay
...
-const ENCODED = [105, 103, 111, 104, 110]   // encoded string "hello"
+const mask = [1, 2, 3, 4]; // 4-byte mask
+const encoded = [105, 103, 111, 104, 110]; // encoded string "hello"
 
-const DECODED = []                          // byte array of decoded payload
-for (let i = 0; i < ENCODED.length; i++) {
-  DECODED[i] = ENCODED[i] ^ MASK[i % 4]
-}
+// Create the byte Array of decoded payload
+const decoded = encoded.from((elt, i) => elt ^ mask[i % 4]); // Perform an XOR on the mask
```
Comment (Josh-Cena) [inline files/en-us/web/api/websockets_api/writing_websocket_servers/index.md:140]: This is actually pseudocode where all-caps seem more idiomatic. Related: https://github.com/mdn/content/issues/18196 I think we should just make it actually usable JS. Also need to update the `**DECODED**` below.

Also, pretty sure it's `%SomeTypedArray%.from(encoded, ...)`?

File: files/en-us/web/api/webvr_api/using_vr_controllers_with_webvr/index.md [hunk 5]
```diff
@@ -66,20 +66,20 @@ Here we first use a tracking variable, `initialRun`, to note that this is the fi
 function reportDisplays() {
   navigator.getVRDisplays().then((displays) => {
     console.log(`${displays.length} displays`);
-    for (let i = 0; i < displays.length; i++) {
-      const cap = displays[i].capabilities;
+    displays.forEach((display, i) => {
+      const cap = display.capabilities;
       // cap is a VRDisplayCapabilities object
       const listItem = document.createElement('li');
       listItem.innerHTML = `<strong>Display ${i+1}</strong><br>` +
```
Comment (Josh-Cena) [inline files/en-us/web/api/webvr_api/using_vr_controllers_with_webvr/index.md:73]:
```suggestion
      listItem.innerHTML = `<strong>Display ${i + 1}</strong><br>` +
```

File: files/en-us/web/api/webxr_device_api/inputs/index.md [hunk 6]
```diff
@@ -207,12 +207,12 @@ The most direct way to decide which controller is primary is to have a user-defi
...
-for (let i=0; i < xrSession.inputSources.length; i++) {
-  if (xrSession.inputSources[i].handedness === user.handedness) {
-    primaryInputSource = inputSources[i];
+xrSession.inputSources.forEach((inputSource) => {
+  if (inputSource.handedness === user.handedness) {
+    primaryInputSource = inputSource;
     break;
   }
-}
+});
```
Comment (Josh-Cena) [inline files/en-us/web/api/webxr_device_api/inputs/index.md:215]:
```js
const primaryInputSource =
  xrSession.inputSources.find((src) => src.handedness === user.handedness) ??
  xrSession.inputSources[0];
```

File: files/en-us/web/api/window/speechsynthesis/index.md [hunk 7]
```diff
@@ -59,9 +59,9 @@ inputForm.onsubmit = (event) => {
...
   const utterThis = new SpeechSynthesisUtterance(inputTxt.value);
   const selectedOption = voiceSelect.selectedOptions[0].getAttribute('data-name');
-  for (let i = 0; i < voices.length ; i++) {
-    if (voices[i].name === selectedOption) {
-      utterThis.voice = voices[i];
+  for (const voice of voices) {
+    if (voice.name === selectedOption) {
+      utterThis.voice = voice;
     }
   }
```
Comment (Josh-Cena) [inline files/en-us/web/api/window/speechsynthesis/index.md:66]:
```suggestion
  utterThis.voice = voices.find((v) => v.name === selectedOption);
```

File: files/en-us/web/api/xmlhttprequest/sending_and_receiving_binary_data/index.md [hunk 8]
```diff
@@ -100,9 +100,9 @@ const array = new ArrayBuffer(512);
 const longInt8View = new Uint8Array(array);
 
 // generate some data
-for (let i=0; i< longInt8View.length; i++) {
-  longInt8View[i] = i % 256;
-}
+longInt8View.forEach((element, i) =>  {
+  element = i % 256;
+});
```
Comment (Josh-Cena) [inline files/en-us/web/api/xmlhttprequest/sending_and_receiving_binary_data/index.md:105]: This needs to be `Uint8Array.from()`

**Review by Josh-Cena — CHANGES_REQUESTED** [review]

File: files/en-us/web/api/xmlhttprequest/sending_and_receiving_binary_data/index.md [hunk 9]
```diff
@@ -97,12 +97,10 @@ You can send JavaScript typed arrays as binary data as well.
...
-const longInt8View = new Uint8Array(array);
-
-// generate some data
-for (let i=0; i< longInt8View.length; i++) {
-  longInt8View[i] = i % 256;
-}
+// Create a new array with fake data (Consecutive numbers (0 - 255), looping back to 0) 
+const longInt8View = Uint8Array.from(array, (element, i) =>  {
+  element = i % 256;
+});
```
Comment (Josh-Cena) [inline files/en-us/web/api/xmlhttprequest/sending_and_receiving_binary_data/index.md:103]:
```suggestion
const longInt8View = Uint8Array.from(array, (element, i) => i % 256);
```

File: files/en-us/web/api/webrtc_api/signaling_and_video_calling/index.md [hunk 10]
```diff
@@ -45,14 +45,8 @@ First up is the addition of the function `sendToOneUser()`. As the name suggests
 ```js
 function sendToOneUser(target, msgString) {
   const isUnique = true;
```
[the remark this answers is not in the document]
Comment (teoli2003) [reply files/en-us/web/api/webrtc_api/signaling_and_video_calling/index.md]: I hesitated, but you are right. It isn't used, so let's get rid of it.

File: files/en-us/web/api/web_audio_api/using_audioworklet/index.md [hunk 11]
```diff
@@ -127,13 +127,10 @@ process(inputList, outputList, parameters) {
     for (let channelNum = 0; channelNum < channelCount; channelNum++) {
       let sampleCount = input[channelNum].length;
```
[the remark this answers is not in the document]
Comment (teoli2003) [reply files/en-us/web/api/web_audio_api/using_audioworklet/index.md:128]: Yes, removed

File: files/en-us/web/api/websockets_api/writing_websocket_servers/index.md [hunk 12]
```diff
@@ -128,18 +128,16 @@ To read the payload data, you must know when to stop reading. That's why the pay
...
 
-### Reading and Unmasking the Data
+### Reading and unmasking the Data
 
-If the MASK bit was set (and it should be, for client-to-server messages), read the next 4 octets (32 bits); this is the masking key. Once the payload length and masking key is decoded, you can read that number of bytes from the socket. Let's call the data **ENCODED**, and the key **MASK**. To get **DECODED**, loop through the octets (bytes a.k.a. characters for text data) of **ENCODED** and XOR the octet with the (i modulo 4)th octet of MASK. In pseudo-code (that happens to be valid JavaScript):
+If the MASK bit was set (and it should be, for client-to-server messages), read the next 4 octets (32 bits); this is the masking key. Once the payload length and masking key is decoded, you can read that number of bytes from the socket. Let's call the data `ENCODED`, and the key `MASK`. To get `DECODED`, loop through the octets (bytes a.k.a. characters for text data) of `encoded` and XOR the octet with the (i modulo 4)th octet of `MASK`. In pseudo-code (that happens to be valid JavaScript):
```
Comment (Josh-Cena) [inline files/en-us/web/api/websockets_api/writing_websocket_servers/index.md:133]:
```suggestion
If the MASK bit was set (and it should be, for client-to-server messages), read the next 4 octets (32 bits); this is the masking key. Once the payload length and masking key is decoded, you can read that number of bytes from the socket. Let's call the data `ENCODED`, and the key `MASK`. To get `DECODED`, loop through the octets (bytes a.k.a. characters for text data) of `ENCODED` and XOR the octet with the (i modulo 4)th octet of `MASK`. In pseudo-code (that happens to be valid JavaScript):
```

File: files/en-us/web/api/websockets_api/writing_websocket_servers/index.md [hunk 13]
```diff
@@ -128,18 +128,16 @@ To read the payload data, you must know when to stop reading. That's why the pay
 2. Read the next 16 bits and interpret those as an unsigned integer. You're **done**.
 3. Read the next 64 bits and interpret those as an unsigned integer. (The most significant bit _must_ be 0.) You're **done**.
 
-### Reading and Unmasking the Data
+### Reading and unmasking the Data
```
Comment (teoli2003) [inline files/en-us/web/api/websockets_api/writing_websocket_servers/index.md:131]:
```suggestion
### Reading and unmasking the data
```

File: files/en-us/web/api/websockets_api/writing_websocket_servers/index.md (hunk 12 above)
Comment (teoli2003) [reply files/en-us/web/api/websockets_api/writing_websocket_servers/index.md:133]: yeah, I didn't have the impression it was pseudo-code from the start.

File: files/en-us/web/api/web_audio_api/using_audioworklet/index.md [hunk 14]
```diff
@@ -156,9 +151,7 @@ process(inputList, outputList, parameters) {
     let channelCount = Math.min(input.length, output.length);
 
     for (let channelNum = 0; channelNum < channelCount; channelNum++) {
-      let sampleCount = input[channelNum].length;
-
-      for (let i = 0; i < sampleCount; i++) {
+      for (let i = 0; i < input[channelNum].lengt; i++) {
```
Comment (Josh-Cena) [inline files/en-us/web/api/web_audio_api/using_audioworklet/index.md:154]:
```suggestion
      for (let i = 0; i < input[channelNum].length; i++) {
```

**Review by Josh-Cena — APPROVED** [review]
LGTM

**Pull request merged**